onnxruntime/__init__.py

```diff
@@ -76,3 +76,80 @@
     __version__ = version
 
 onnxruntime_validation.check_distro_info()
+
+# Load nvidia libraries from site-packages/nvidia if the package is onnxruntime-gpu
+if (
+    __package__ == "onnxruntime-gpu"
+    # Just in case we rename the package name in the future
+    or __package__ == "onnxruntime-cuda"
+    or __package__ == "onnxruntime_gpu"
+    or __package__ == "onnxruntime_cuda"
+):
+    import ctypes
+    import os
+    import platform
+    import re
+    import site
+    import logging
+    # Get the site-packages path where nvidia packages are installed
+    site_packages_path = site.getsitepackages()[-1]
+    nvidia_path = os.path.join(site_packages_path, "nvidia")
+    # Traverse the directory and subdirectories
+    if platform.system() == "Windows":  #
+        # Define the list of DLL patterns, curand and nvJitLink are not included for Windows
+        cuda_libs = (
+            "cublas",
+            "cublasLt",
+            "cudnn",
+            "cudart",
+            "cufft",
+            # "curand",
+            # "nvJitLink",
+        )
+        # Construct a regex pattern for each library name with optional parts
+        # Pattern explanation:
+        # - `libname`: Match the base library name (e.g., "cudart")
+        # - `(?:64)?`: Optionally match "64"
+        # - `(?:_\d+)*`: Match zero or more occurrences of "_n" where "n" is one or more digits
+        # - `.dll$`: End with ".dll" ignoring case
+        lib_pattern = {lib: re.compile(rf"{lib}(?:64)?(?:_\d+)*\.dll$", re.IGNORECASE) for lib in cuda_libs}
+        # Collect all directories under site-packages/nvidia that contain .dll files (for Windows)
+        for root, _, files in os.walk(nvidia_path):
+            # Add the current directory to the DLL search path
+
+            with os.add_dll_directory(root):
+                # Find all .dll files in the current directory
+                for file in files:
+                    for pattern in lib_pattern.items().values():
+                        if pattern.match(file):
+                            dll_path = os.path.join(root, file)
+                            try:
+                                _ = ctypes.CDLL(dll_path)
+                            except Exception as e:
+                                logging.error(f"Failed to load {dll_path}: {e}")
+    elif platform.system() == "Linux":
+        # Define the patterns with optional version number and case-insensitivity
+        cuda_libs = (
+            "libcublas.so",
+            "libcublasLt.so",
+            "libcudnn.so",
+            "libcudart.so",
+            "libnvrtc.so",
+            "libcufft.so",
+            "libcurand.so",
+            "libnvJitLink.so",
+        )
+
+        # Regular expression to match .so files with optional versioning (e.g., .so, .so.1, .so.2.3)
+        lib_pattern = {pattern: re.compile(rf"{re.escape(pattern)}(\.\d+)*$", re.IGNORECASE) for pattern in cuda_libs}
```

Comment:
Suggest to use a list of fixed filename instead of regular expression.

We shall not assume that user install the correct version in site package. For example, user could install cudnn 8 in site package (for other application), but also install cudnn 9 (not in site package). Then we shall not load the cudnn 8 in site-package.